src/agent/debugger/src/target.rs

```diff
@@ -479,17 +482,83 @@ impl Target {
     }
 
     pub fn prepare_to_resume(&mut self) -> Result<()> {
-        if let Some(context) = self.current_context.take() {
-            if self.context_is_modified {
-                context.set_thread_context(self.current_thread_handle)?;
+        // When resuming, we take extra care to avoid missing breakpoints. There are 2 race
+        // conditions considered here:
+        //
+        // * Threads A and B hit a breakpoint at literally the same time, so we have
+        //   multiple events already queued to handle those breakpoints.
+        //   Restoring the breakpoint for thread A should not interfere with restoring
+        //   the original code for thread B.
+        // * Thread A hit a breakpoint, thread B is 1 instruction away from hitting
+        //   that breakpoint. Thread B cannot miss the breakpoint when thread A is resumed.
+        //
+        // To avoid these possible races, when resuming, we only let a single thread go **if**
```

Comment:
Are there any integration/regression tests we could add that would trigger the race in the old impl, with high probability?

Reply:
I'll give this more thought. I'm also hitting a real deadlock caused by this new logic, so it would be good to get a test that can trigger these sorts of issues.